flink-core/src/main/java/org/apache/flink/api/java/typeutils/FieldAccessor.java

```diff
@@ -0,0 +1,295 @@
+			Tuple tuple = (Tuple) record;
+			tuple.setField(fieldValue, pos);
+			return record;
+		}
+	}
+
+	/**
+	 * @param <T> The Tuple type
+	 * @param <R> The field type at the first level
+	 * @param <F> The field type at the innermost level
+	 */
+	static final class RecursiveTupleFieldAccessor<T, R, F> extends FieldAccessor<T, F> {
+
+		private static final long serialVersionUID = 1L;
+
+		private final int pos;
+		private final FieldAccessor<R, F> innerAccessor;
+
+		RecursiveTupleFieldAccessor(int pos, FieldAccessor<R, F> innerAccessor) {
+			this.pos = pos;
+			this.innerAccessor = innerAccessor;
+			this.fieldType = innerAccessor.fieldType;
+		}
+
+		@Override
+		public F get(T record) {
+			Tuple tuple = (Tuple) record;
+			final R inner = tuple.getField(pos);
+			return innerAccessor.get(inner);
+		}
+
+		@Override
+		public T set(T record, F fieldValue) {
+			Tuple tuple = (Tuple) record;
+			final R inner = tuple.getField(pos);
+			tuple.setField(innerAccessor.set(inner, fieldValue), pos);
+			return record;
+		}
+	}
+
+	/**
+	 * @param <T> The POJO type
+	 * @param <R> The field type at the first level
+	 * @param <F> The field type at the innermost level
+	 */
+	static final class PojoFieldAccessor<T, R, F> extends FieldAccessor<T, F> {
+
+		private static final long serialVersionUID = 1L;
+
+		private transient Field field;
+		private final FieldAccessor<R, F> innerAccessor;
+
+		PojoFieldAccessor(Field field, FieldAccessor<R, F> innerAccessor) {
+			this.field = field;
+			this.innerAccessor = innerAccessor;
+			this.fieldType = innerAccessor.fieldType;
+		}
+
+		@Override
+		public F get(T pojo) {
+			try {
+				@SuppressWarnings("unchecked")
+				final R inner = (R)field.get(pojo);
+				return innerAccessor.get(inner);
+			} catch (IllegalAccessException iaex) {
+				throw new RuntimeException("This should not happen since we call setAccesssible(true) in readObject."
+						+ " fields: " + field + " obj: " + pojo);
+			}
+		}
+
+		@Override
+		public T set(T pojo, F valueToSet) {
+			try {
+				@SuppressWarnings("unchecked")
+				final R inner = (R)field.get(pojo);
+				field.set(pojo, innerAccessor.set(inner, valueToSet));
+				return pojo;
+			} catch (IllegalAccessException iaex) {
+				throw new RuntimeException("This should not happen since we call setAccesssible(true) in readObject."
+						+ " fields: " + field + " obj: " + pojo);
+			}
+		}
+
+		private void writeObject(ObjectOutputStream out)
+				throws IOException, ClassNotFoundException {
+			out.defaultWriteObject();
+			FieldSerializer.serializeField(field, out);
+		}
+
+		private void readObject(ObjectInputStream in)
+				throws IOException, ClassNotFoundException {
+			in.defaultReadObject();
+			field = FieldSerializer.deserializeField(in);
+		}
+	}
+
+
+	// ==================================================================================================
+
+
+	private final static String REGEX_FIELD = "[\\p{L}_\\$][\\p{L}\\p{Digit}_\\$]*";
```

Comment:
Maybe we can move those regex strings to `ExpressionKeys` and remove the duplicate code in `CaseClassTypeInfo`.